flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SystemProcessingTimeService.java

```diff
@@ -253,89 +243,53 @@ int getNumTasksScheduled() {
-		public void run() {
-			synchronized (lock) {
-				try {
-					if (serviceStatus.get() == STATUS_ALIVE) {
-						target.onProcessingTime(timestamp);
-					}
-				} catch (Throwable t) {
-					TimerException asyncException = new TimerException(t);
-					exceptionHandler.handleAsyncException("Caught exception while processing timer.", asyncException);
-				}
-			}
-		}
+	private Runnable wrapOnTimerCallback(ProcessingTimeCallback callback, long timestamp) {
+		return new TimeTrackingDelegate(status, callbackExecutionContext, callback, timestamp, 0);
 	}
 
-	/**
-	 * Internal task which is repeatedly called by the processing time service.
-	 */
-	private static final class RepeatedTriggerTask implements Runnable {
+	private Runnable wrapOnTimerCallback(ProcessingTimeCallback callback, long nextTimestamp, long period) {
+		return new TimeTrackingDelegate(status, callbackExecutionContext, callback, nextTimestamp, period);
+	}
 
+	private static final class TimeTrackingDelegate implements Runnable {
 		private final AtomicInteger serviceStatus;
-		private final Object lock;
-		private final ProcessingTimeCallback target;
-		private final long period;
-		private final AsyncExceptionHandler exceptionHandler;
+		private final ScheduledCallbackExecutionContext callbackExecutionContext;
+		private final ProcessingTimeCallback callback;
 
 		private long nextTimestamp;
+		private final long period;
 
-		private RepeatedTriggerTask(
-				final AtomicInteger serviceStatus,
-				final AsyncExceptionHandler exceptionHandler,
-				final Object lock,
-				final ProcessingTimeCallback target,
-				final long nextTimestamp,
-				final long period) {
-
-			this.serviceStatus = Preconditions.checkNotNull(serviceStatus);
-			this.lock = Preconditions.checkNotNull(lock);
-			this.target = Preconditions.checkNotNull(target);
+		TimeTrackingDelegate(
+				AtomicInteger serviceStatus,
+				ScheduledCallbackExecutionContext callbackExecutionContext,
+				ProcessingTimeCallback callback,
+				long timestamp,
+				long period) {
+			this.serviceStatus = serviceStatus;
+			this.callbackExecutionContext = callbackExecutionContext;
+			this.callback = callback;
+			this.nextTimestamp = timestamp;
 			this.period = period;
-			this.exceptionHandler = Preconditions.checkNotNull(exceptionHandler);
-
-			this.nextTimestamp = nextTimestamp;
 		}
 
 		@Override
 		public void run() {
-			synchronized (lock) {
-				try {
-					if (serviceStatus.get() == STATUS_ALIVE) {
-						target.onProcessingTime(nextTimestamp);
-					}
-
-					nextTimestamp += period;
-				} catch (Throwable t) {
-					TimerException asyncException = new TimerException(t);
-					exceptionHandler.handleAsyncException("Caught exception while processing repeated timer task.", asyncException);
-				}
+			if (serviceStatus.get() != STATUS_ALIVE) {
+				return;
+			}
+			try {
+				callbackExecutionContext.invoke(callback, nextTimestamp);
+			} catch (InterruptedException ex) {
+				Thread.currentThread().interrupt();
```

Comment:
(+ to the other your comment about `InterruptedException`)
Frankly, I also wonder how and where to handle this exception.
Same question applies for checkpoints handling in the main task thread (FLINK-12482).

Some of my thoughts:

The root of this exception is in `ReentrantLock.lockInterruptibly()` used in the mailbox.
If the exception is actually thrown, I interpret it that the mailbox queue is full (the sender thread had to sleep, assumption that the threads in the time service thread pool don't have interrupted status set). So if something external interrupts the time service thread on mailbox, I assume it's somethings "disastrous", like forced shutdown of task manager.

Here, I handle the exception in the `SystemProcessingTimeService`, as the thread belongs to the thread pool of this service. So the service may decide what to do with such exception.

On the time service side, basically there is not much to do with it and the thread actually was "told" to be interrupted. So for now, the `Thread.currentThread().interrupt();` action is the reaction to such situation.

Reply:
First, the exception has not really something to do with the question if the queue is full, but rather if the thread that calls the method was interrupted. Now we can ask ourselves if there are currenlty parts in the code that could interrupt this particular thread that wants to enqueue timers. I think this could happen if the threadpool to which it belongs is shut down.

Now the next question is, under what circumstances can this interruption happen and what is the consequence for the thread? Maybe the answer is: it only happens when the task is ending (for whatever reason) and then there might be nothing left to do. Or are there any other reasons, then the reaction might be that we have to drop a letter and for correctness this is the time to tell the task that it should fail. For scenario one, your approach looks ok, but if scneario 2 applies, the thread should also forward this as async exception so that the job can fail, and this is more easily done inside the method and not on the caller side.